subsys/bluetooth/shell/bt.c

```diff
@@ -1245,6 +1247,292 @@ static int cmd_adv_info(const struct shell *shell, size_t argc, char *argv[])
+		return -EINVAL;
+	}
+
+	ad_len = hex2bin(argv[1], strlen(argv[1]), hex_data, sizeof(hex_data));
+
+	if (!ad_len) {
+		shell_error(shell, "Could not parse adv data");
+		return -ENOEXEC;
+	}
+
+	ad.data_len = hex_data[0];
+	ad.type = hex_data[1];
+	ad.data = &hex_data[2];
+
+	err = bt_le_per_adv_set_data(adv, &ad, ad_len);
+	if (err) {
+		shell_error(shell,
+			    "Failed to set periodic advertising data (%d)",
+			    err);
+		return -ENOEXEC;
+	}
+
+	return 0;
+}
+#endif /* CONFIG_BT_PER_ADV */
+
+#if defined(CONFIG_BT_PER_ADV_SYNC)
+static struct bt_le_per_adv_sync *per_adv_syncs[CONFIG_BT_PER_ADV_SYNC_MAX];
+static struct bt_le_per_adv_sync *selected_per_adv_sync;
+
+static void per_adv_sync_sync_cb(struct bt_le_per_adv_sync *sync,
+				 struct bt_le_per_adv_sync_synced_info *info)
+{
+	char le_addr[BT_ADDR_LE_STR_LEN];
+
+	if (selected_per_adv_sync != sync) {
+		shell_error(ctx_shell, "Invalid per adv sync object");
+		return;
+	}
+
+	bt_addr_le_to_str(info->addr, le_addr, sizeof(le_addr));
+	shell_print(ctx_shell, "PER_ADV_SYNC[%u]: [DEVICE]: %s synced, "
+		    "Interval 0x%04x (%u ms), PHY %s",
+		    bt_le_per_adv_sync_get_index(sync), le_addr,
+		    info->interval, info->interval * 5 / 4, phy2str(info->phy));
+}
+
+static void per_adv_sync_terminated_cb(
+	struct bt_le_per_adv_sync *sync,
+	const struct bt_le_per_adv_sync_term_info *info)
+{
+	char le_addr[BT_ADDR_LE_STR_LEN];
+
+	if (selected_per_adv_sync != sync) {
+		shell_error(ctx_shell, "Invalid per adv sync object");
+		return;
+	}
+
+	bt_addr_le_to_str(info->addr, le_addr, sizeof(le_addr));
+	shell_print(ctx_shell, "PER_ADV_SYNC[%u]: [DEVICE]: %s sync terminated",
+		    bt_le_per_adv_sync_get_index(sync), le_addr);
+	selected_per_adv_sync = NULL;
+}
+
+static void per_adv_sync_recv_cb(
+	struct bt_le_per_adv_sync *sync,
+	const struct bt_le_per_adv_sync_recv_info *info,
+	struct net_buf_simple *buf)
+{
+	char le_addr[BT_ADDR_LE_STR_LEN];
+
+	if (selected_per_adv_sync != sync) {
+		shell_error(ctx_shell, "Invalid per adv sync object");
+		return;
+	}
+
+	bt_addr_le_to_str(info->addr, le_addr, sizeof(le_addr));
+	shell_print(ctx_shell, "PER_ADV_SYNC[%u]: [DEVICE]: %s, tx_power %i, "
+		    "RSSI %i, CTE %u, data length %u",
+		    bt_le_per_adv_sync_get_index(sync), le_addr, info->tx_power,
+		    info->rssi, info->cte_type, buf->len);
+}
+
+static struct bt_le_per_adv_sync_cb per_adv_sync_cb = {
+	.synced = per_adv_sync_sync_cb,
+	.term = per_adv_sync_terminated_cb,
+	.recv = per_adv_sync_recv_cb
+};
+
+static int cmd_per_adv_sync_create(const struct shell *shell, size_t argc,
+				   char *argv[])
+{
+	int err;
+	struct bt_le_per_adv_sync_param create_params = { 0 };
+	uint32_t options = 0;
+	struct bt_le_per_adv_sync *free_per_adv_sync = NULL;
+
+	for (int i = 0; i < ARRAY_SIZE(per_adv_syncs); i++) {
+		if (per_adv_syncs[i] == NULL) {
+			free_per_adv_sync = per_adv_syncs[i];
+			break;
+		}
+	}
+
+	if (!free_per_adv_sync) {
```

Comment:
`free_per_adv_sync` is always `NULL` as `free_per_adv_sync = per_adv_syncs[i];`

Reply:
If all pointers in `per_adv_syncs` are non-NULL, then `free_per_adv_sync` will also be NULL. I think this is correct.

Reply:
check line 1439, you break on first NULL and NULL is assigned to `free_per_adv_sync`. Has this shell command tested to work? I always get "Cannot create more per adv syncs".

Reply:
Ah, yes I now see what you mean. I did change it recently, and actually didn't test the shell command afterwards. I changed the check now to something that works.